test/e2e/apimachinery/apply.go

```diff
@@ -0,0 +1,303 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	"github.com/onsi/ginkgo"
+
+	// ensure libs have a chance to initialize
+	_ "github.com/stretchr/testify/assert"
+)
+
+var _ = SIGDescribe("ServerSideApply", func() {
+	f := framework.NewDefaultFramework("apply")
+
+	var client clientset.Interface
+	var ns string
+
+	ginkgo.BeforeEach(func() {
+		client = f.ClientSet
+		ns = f.Namespace.Name
+	})
+
+	ginkgo.AfterEach(func() {
+	})
+
+	ginkgo.It("PATCH requests with the apply content type will create the object if it does not already exist", func() {
+		{
+			testCases := []struct {
+				resource string
+				name     string
+				body     string
+			}{
+				{
+					resource: "pods",
+					name:     "test-pod",
+					body: `{
+				"apiVersion": "v1",
+				"kind": "Pod",
+				"metadata": {
+					"name": "test-pod"
+				},
+				"spec": {
+					"containers": [{
+						"name":  "test-container",
+						"image": "test-image"
+					}]
+				}
+			}`,
+				}, {
+					resource: "services",
+					name:     "test-svc",
+					body: `{
+				"apiVersion": "v1",
+				"kind": "Service",
+				"metadata": {
+					"name": "test-svc"
+				},
+				"spec": {
+					"ports": [{
+						"port": 8080,
+						"protocol": "UDP"
+					}]
+				}
+			}`,
+				},
+			}
+
+			for _, tc := range testCases {
+				_, err := client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
+					Namespace(ns).
+					Resource(tc.resource).
+					Name(tc.name).
+					Param("fieldManager", "apply_test").
+					Body([]byte(tc.body)).
+					Do(context.TODO()).
+					Get()
+				if err != nil {
+					framework.Failf("Failed to create object using Apply patch: %v", err)
+				}
+
+				_, err = client.CoreV1().RESTClient().Get().Namespace(ns).Resource(tc.resource).Name(tc.name).Do(context.TODO()).Get()
+				if err != nil {
+					framework.Failf("Failed to retrieve object: %v", err)
+				}
+
+				// Test that we can re apply with a different field manager and don't get conflicts
+				_, err = client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
+					Namespace(ns).
+					Resource(tc.resource).
+					Name(tc.name).
+					Param("fieldManager", "apply_test_2").
+					Body([]byte(tc.body)).
+					Do(context.TODO()).
+					Get()
+				if err != nil {
+					framework.Failf("Failed to re-apply object using Apply patch: %v", err)
+				}
+			}
+		}
+	})
+
+	ginkgo.It("PUT requests which change nothing should not change the resource version", func() {
```

Comment:
we have TestNoOpUpdateSameResourceVersion as an integration test... do we need an e2e as well?

Reply:
Yeah I agree the two tests should either both be in e2e or both in integration, it doesn't make sense to move just one of them. Moved this one outside this PR for now, can add it back in later if we need.